### PR TITLE
[exporter/signalfxexporter] Add all HTTP client settings to the configuration

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Expose all HTTP client settings to the exporter configuration so they may be set in addition to the existing flags.

**Link to tracking Issue:**
#16807

**Testing:**
Unit tests all pass.

**Documentation:**
N/A - not sure we need to shore up those config settings explicitly as I haven't seen that done elsewhere - please advise.

### CODE REVIEW

File: exporter/signalfxexporter/exporter.go [hunk 1]
```diff
@@ -175,47 +194,62 @@ func newGzipPool() sync.Pool {
...
 }
 
-func newEventExporter(config *Config, logger *zap.Logger) (*signalfxExporter, error) {
+func newEventExporter(config *Config, createSettings exporter.CreateSettings) (*signalfxExporter, error) {
 	if config == nil {
 		return nil, errors.New("nil config")
 	}
 
-	options, err := config.getOptionsFromConfig()
+	_, err := config.getOptionsFromConfig()
```
Comment (MovieStoreGuy) [inline exporter/signalfxexporter/exporter.go:202]: Is this required anymore?

I imagine if there is an issue processing the config, it should be caught in `config.Validate()`

File: exporter/signalfxexporter/exporter.go [hunk 2]
```diff
@@ -230,3 +264,24 @@ func (se *signalfxExporter) pushLogs(ctx context.Context, ld plog.Logs) error {
 	_, err := se.pushLogsData(ctx, ld)
 	return err
 }
+
+func buildHeaders(config *Config) map[string]string {
+	headers := map[string]string{
+		"Connection":   "keep-alive",
+		"Content-Type": "application/x-protobuf",
+		"User-Agent":   "OpenTelemetry-Collector SignalFx Exporter/v0.0.1",
```
Comment (MovieStoreGuy) [inline exporter/signalfxexporter/exporter.go:272]: I wonder if there is a means to get the published collector version to help capture analytics within SignalFx?

**Review by MovieStoreGuy — COMMENTED** [review]
Just a couple of comments but I do appreciate making the change so quickly

File: exporter/signalfxexporter/exporter.go (hunk 1 above)
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:202]: it's a bit contrived, but I wanted to make this without changing the tests too much. Agree we can refactor more, but that felt like a bit more work than what I was ready to deliver.

File: exporter/signalfxexporter/exporter.go (hunk 2 above)
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:272]: Totally. Actually, I think that could be done via an extension or using the host BuildInfo. Again, felt like a bridge too far for those changes. It is worth filing an issue to follow up imo.

Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:272]: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16841 is filed to follow up.

File: exporter/signalfxexporter/exporter.go [hunk 3]
```diff
@@ -106,36 +108,55 @@ func newSignalFxExporter(
...
+}
+
+func (se *signalfxExporter) start(_ context.Context, host component.Host) (err error) {
+	options, err := se.config.getOptionsFromConfig()
+	if err != nil {
+		return err
+	}
+
+	headers := buildHeaders(se.config)
+	se.config.Headers = nil
```
Comment (fatsheep9146) [inline exporter/signalfxexporter/exporter.go:126]: Line 127 - Line 137  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16837/files#diff-8263390249c86b6a1730e8dc0732629a1c7e707797f96ad5745a24e11d284f85R127-R137
seems same as 
Line 222 - Line 234 https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16837/files#diff-8263390249c86b6a1730e8dc0732629a1c7e707797f96ad5745a24e11d284f85R222-R234

Maybe convert those logic into a common function is better?

Comment (atoulme) [reply exporter/signalfxexporter/exporter.go]: Good idea, done!

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
I am fine with these changes, however, @dmitryax and @pmcollins should verify these.

**Review by fatsheep9146 — APPROVED** [review]

File: exporter/signalfxexporter/exporter.go [hunk 4]
```diff
@@ -175,47 +181,66 @@ func newGzipPool() sync.Pool {
...
+	return nil
+}
+
+func (se *signalfxExporter) createClient(host component.Host) (*http.Client, error) {
+
+	se.config.HTTPClientSettings.Headers = nil
+	se.config.HTTPClientSettings.TLSSetting = se.config.IngestTLSSettings
+
+	if se.config.MaxIdleConns == nil {
+		se.config.MaxIdleConns = &se.config.MaxConnections
```
[the remark this answers is not in the document]
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:233]: We should deprecate it, but I wanted to hear from you on the best route to do so. The reason for withholding my decision is that `HTTPClientSettings` settings are not documented on the receiver config at this time. I think at the minimum we must explicitly document those if we deprecate `MaxConnections`. Is there a proper way to document `HTTPClientSettings` that I can use?

File: exporter/signalfxexporter/exporter.go [hunk 5]
```diff
@@ -175,47 +181,66 @@ func newGzipPool() sync.Pool {
...
+	}
+	if se.config.MaxIdleConnsPerHost == nil {
```
[the remark this answers is not in the document]
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:235]: oh ok. I'll take your guidance on this.

File: exporter/signalfxexporter/exporter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/exporter.go:233]: Yes, we need to document it by adding a link to https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confighttp. similar to how it's done for other exporters https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/dynatraceexporter#advanced-configuration

[the remark this answers is not in the document]
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:233]: Great, I'll work on that.

[the remark this answers is not in the document]
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go]: added links

File: exporter/signalfxexporter/config.go [hunk 6]
```diff
@@ -111,6 +111,7 @@ type Config struct {
 	NonAlphanumericDimensionChars string `mapstructure:"nonalphanumeric_dimension_chars"`
 
 	// MaxConnections is used to set a limit to the maximum idle HTTP connection the exporter can keep open.
+	// Deprecated: prefer using HTTPClientSettings.MaxIdleConns or HTTPClientSettings.MaxIdleConns instead.
```
Comment (dmitryax) [inline exporter/signalfxexporter/config.go:114]:
```suggestion
	// Deprecated: prefer using HTTPClientSettings.MaxIdleConns or HTTPClientSettings.MaxIdleConnsPerHost instead.
```

File: exporter/signalfxexporter/exporter.go [hunk 7]
```diff
@@ -175,47 +182,70 @@ func newGzipPool() sync.Pool {
...
+
+	if se.config.HTTPClientSettings.MaxIdleConns == nil {
+		se.config.HTTPClientSettings.MaxIdleConns = &se.config.MaxConnections
+	}
+	if se.config.HTTPClientSettings.MaxIdleConnsPerHost == nil {
+		se.config.HTTPClientSettings.MaxIdleConnsPerHost = &se.config.MaxConnections
+	}
+	if se.config.HTTPClientSettings.MaxConnsPerHost == nil {
+		se.config.HTTPClientSettings.MaxConnsPerHost = &se.config.MaxConnections
+	}
```
Comment (dmitryax) [inline exporter/signalfxexporter/exporter.go:242]: Why is this needed? We didn't touch set `MaxConnsPerHost` before, right?

File: exporter/signalfxexporter/exporter.go [hunk 8]
```diff
@@ -175,47 +182,70 @@ func newGzipPool() sync.Pool {
...
-		pushLogsData: eventClient.pushLogsData,
-	}, nil
+	se.pushLogsData = eventClient.pushLogsData
+	return nil
+}
+
+func (se *signalfxExporter) createClient(host component.Host) (*http.Client, error) {
+
+	// we want to control how headers are set, overriding user headers with our passthrough.
+	se.config.HTTPClientSettings.Headers = nil
```
Comment (dmitryax) [inline exporter/signalfxexporter/exporter.go:231]: We should not mutate the original config. Let's make a copy first

File: exporter/signalfxexporter/README.md [hunk 9]
```diff
@@ -89,7 +89,7 @@ The following configuration options can also be configured:
 that are allowed to be used as a dimension key in addition to alphanumeric 
 characters. Each nonalphanumeric dimension key character that isn't in this string 
 will be replaced with a `_`.
-- `max_connections` (default = 100):  The maximum number of idle HTTP connection the exporter can keep open.
+- `max_connections` (default = 100):  The maximum number of idle HTTP connection the exporter can keep open. Deprecated. Prefer using [HTTP settings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/README.md) instead.
```
Comment (dmitryax) [inline exporter/signalfxexporter/README.md:92]: I would rather reword it to insist on not using it. We will need to remove deprecated fields eventually.
```suggestion
- `max_connections` (default = 100):  The maximum number of idle HTTP connections the exporter can keep open. Deprecated: use `max_idle_conns` or `max_idle_conns_per_host` instead. See [HTTP settings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/README.md) for more info.
```

File: exporter/signalfxexporter/exporter.go (hunk 7 above)
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go:242]: I assumed that was needed and missing before. I'll remove this addition.

File: exporter/signalfxexporter/exporter.go (hunk 8 above)
Comment (atoulme) [reply exporter/signalfxexporter/exporter.go]: We do make a copy first, but before this method is called. The headers are then set - I'm going to move code around to make that more obvious.

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
Ready to merge once @dmitryax or @pmcollins approves.

**Review by dmitryax — APPROVED** [review]
